Change: Implement side-by-side font cache, allowing use of multiple fonts.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Font system only allows one active font at a time, but this means that it may be impossible to find a font that covers all required characters.

Currently if the glyph search detects this, it will force using a font that may not be suitable.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Allow multiple fonts to be configured side-by-side.

If a fallback font is required, the default OpenTTD font will also be loaded along side it.

Multiple fonts can also be configured explicitly.

~~Now rebased on top of #14448, #14449 and #14450~~

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

Does not build for Mac OS as I'm not a Mac OS dev...

Font auto-detection is still not great.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

LordAro commented on src/fontcache.cpp:
```diff
@@ -17,13 +19,14 @@
 /* static */ const int FontCache::DEFAULT_FONT_HEIGHT[FS_END] = {10, 6, 18, 10};
 /** Default unscaled ascenders for the different sizes of fonts. */
-/* static */ const int FontCache::DEFAULT_FONT_ASCENDER[FS_END] = {8, 5, 15, 8};
+/* static */ const int FontCache::DEFAULT_FONT_ASCENDER[FS_END] = {8, 6, 15, 8};
```
this feels like an unrelated change. can you elaborate?

PeterN replied:
With this PR, fonts are aligned vertically by their baseline instead of being centred within the largest bounds. This allows text with a mix of fonts to be rendered correctly.

The original sprite-based small font only contains uppercase letters (repeated for lowercase) and has no descenders, and the baseline is actually the bottom of the sprite, which is what this change is for.

<img width="94" height="85" alt="image" src="https://github.com/user-attachments/assets/c739038b-c2c9-4d84-be6b-8efa3ebe96cb" />

But on the other hand, OpenGFX2 does it differently and moves everything up 1 pixel, which is why switching between sprite font OpenGFX2 and non-sprite font OpenTTD-Small causes small text to be moved.

PeterN committed:
Change: Support side-by-side fallback FontCaches instead of hierarchical.

The text layouter system can now support using different fonts for different glyphs, including mixing scalable and sprite glyphs.

rubidium42 commented on src/fontcache.cpp:
```diff
@@ -52,32 +55,44 @@ FontCacheSettings _fcsettings;
  * @return true if a font has been set, false otherwise.
  */
-/* static */ bool FontProviderManager::FindFallbackFont(FontCacheSettings *settings, const std::string &language_isocode, MissingGlyphSearcher *callback)
+/* static */ bool FontProviderManager::FindFallbackFont(const std::string &language_isocode, FontSizes fontsizes, class MissingGlyphSearcher *callback)
```
```suggestion
/* static */ bool FontProviderManager::FindFallbackFont(const std::string &language_isocode, FontSizes fontsizes, MissingGlyphSearcher *callback)
```
Does this `class` need to be added, `MissingGlyphSearcher` hasn't moved.

rubidium42 commented on src/fontcache.cpp:
```diff
@@ -216,18 +227,54 @@ std::string GetFontCacheFontName(FontSize fs)
+	* @param fontsizes Fontsizes to add fallback to.
+	* @param name Name of font to add.
+	* @param handle OS-specific handle or data of font.
+	*/
```
```suggestion
/**
 * Add a fallback font, with optional OS-specific handle.
 * @param fontsizes Fontsizes to add fallback to.
 * @param name Name of font to add.
 * @param handle OS-specific handle or data of font.
 */
```

rubidium42 commented on src/os/unix/font_unix.cpp:
```diff
@@ -148,22 +152,25 @@ bool FontConfigFindFallbackFont(FontCacheSettings *settings, const std::string &
 	auto fc_instance = AutoRelease<FcConfig, FcConfigDestroy>(FcConfigReference(nullptr));
 	assert(fc_instance != nullptr);
 
+	/* Get set of required characters. XXX Do we know what font size we want here? */
```
Is the `XXX` part still valid? We are passing `fontsizes`.

rubidium42 commented on src/os/unix/font_unix.cpp:
```diff
@@ -185,26 +192,32 @@ bool FontConfigFindFallbackFont(FontCacheSettings *settings, const std::string &
+		}
+
+		if (matching_chars < chars.size()) {
+			Debug(fontcache, 1, "Font \"{}\" misses {} glyphs", (char *)file, chars.size() - matching_chars);
```
```suggestion
			Debug(fontcache, 1, "Font \"{}\" misses {} glyphs", static_cast<char *>(file), chars.size() - matching_chars);
```
